node/lib/util/push.js

```diff
@@ -84,23 +84,75 @@ exports.push = co.wrap(function *(repo, remoteName, source, target, force) {
     let remoteUrl = yield GitUtil.getUrlFromRemoteName(repo, remoteName);
 
     // First, push the submodules.
+    const annotatedCommit = yield GitUtil.resolveCommitish(repo, source);
+    const sha = annotatedCommit.id();
+    const commit = yield repo.getCommit(sha);
+
+    const submoduleSet = new Set(yield SubmoduleUtil.listOpenSubmodules(repo));
+    for (const s of yield SubmoduleUtil.listAbsorbedSubmodules(repo)) {
+        submoduleSet.add(s);
+    }
+    const submodules = [...submoduleSet];
+    const pushMap = yield SubmoduleUtil.getSubmoduleShasForCommit(repo,
+                                                                  submodules,
+                                                                  commit);
+    const trackingBranches = new Set();
+    trackingBranches.add(`refs/remotes/${remoteName}/${target}`);
+    let tracking;
+    try {
+        const sourceBranch = yield repo.getReference(source);
+        tracking = yield GitUtil.getTrackingInfo(repo, sourceBranch);
+    } catch (e) {
+        // we have no local branch? maybe source is a sha.
+        tracking = null;
+    }
+
+    if (tracking !== null) {
+        if (tracking.remoteName !== null) {
+            trackingBranches.add(
+                `refs/remotes/${remoteName}/${tracking.remoteName}`);
+        }
+        if (tracking.pushRemoteName !== null) {
+            trackingBranches.add(
+                `refs/remotes/${remoteName}/${tracking.pushRemoteName}`);
+        }
+    }
+
+    for (const branch of trackingBranches) {
+        const annotated = yield GitUtil.resolveCommitish(repo, branch);
+
+        if (annotated !== null) {
+            const trackingCommit = yield NodeGit.Commit.lookup(repo,
+                                                               annotated.id());
+            const tree = yield trackingCommit.getTree();
+            for (const sub of Object.keys(pushMap)) {
+                let entry;
+                try {
+                    entry = yield tree.entryByPath(sub);
+                } catch (e) {
+                    //not found, OK
+                    continue;
+                }
+                // easy check: does the tracking branch's id equal the one
+                // in the pushMap?  If not, it would be possible that the
+                // tracking branch were *ahead*, but this is so unlikely as
+                // to be not worth opening up the repo to check
+                if (entry.id() === pushMap[sub]) {
```

Comment:
I think you should open up the repo and check in the case that they're not equal. It would be great if you can come up with a way to do this without fully opening repos that are just absorbed -- I'll reuse this logic in places like `merge` where I sometimes have to fully open a repo just to decide that something could have been fast-forwarded.

It seems fairly likely that the SHAs from the upstream branch may have progressed before you've pushed to your remote.